tools/build_langserver/langserver/analyzer.go

```diff
@@ -139,35 +170,99 @@ func (a *Analyzer) IdentFromFile(uri lsp.DocumentURI) ([]*Identifier, error) {
 		return nil, err
 	}
 
-	var idents []*Identifier
+	return stmts, nil
+}
+
+// StatementFromPos returns a Statement struct with either an Identifier or asp.Expression
+func (a *Analyzer) StatementFromPos(uri lsp.DocumentURI, position lsp.Position) (*Statement, error) {
+	// Get all the statements from the build file
+	stmts, err := a.AspStatementFromFile(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	return a.getStatementByPos(stmts, position)
+}
+
+func (a *Analyzer) getStatementByPos(stmts []*asp.Statement, position lsp.Position) (*Statement, error) {
 	for _, stmt := range stmts {
-		if stmt.Ident != nil {
-			// get the identifier type
-			var identType string
-			if stmt.Ident.Action != nil {
-				if stmt.Ident.Action.Property != nil {
-					identType = "property"
-				} else if stmt.Ident.Action.Call != nil {
-					identType = "call"
-				} else if stmt.Ident.Action.Assign != nil {
-					identType = "assign"
-				} else if stmt.Ident.Action.AugAssign != nil {
-					identType = "augAssign"
-				}
+		if withInRange(stmt.Pos, stmt.EndPos, position) {
+			// get function call, assignment, and property access
+			if stmt.Ident != nil {
+				return &Statement{
+					Ident: a.identFromStatement(stmt),
+				}, nil
 			}
+			return a.statementFromAst(reflect.ValueOf(stmt), position)
+		}
+	}
+
+	return nil, nil
+}
 
-			ident := &Identifier{
-				IdentStatement: stmt.Ident,
-				Type:           identType,
-				// -1 from asp.Statement.Pos.Line, as lsp position requires zero index
-				StartLine: stmt.Pos.Line - 1,
-				EndLine:   stmt.EndPos.Line - 1,
+// statementFromAsp recursively finds asp.IdentStatement and asp.Expression in the ast
+// and returns a valid Statement pointer if within range
+func (a *Analyzer) statementFromAst(v reflect.Value, position lsp.Position) (*Statement, error) {
```

Comment:
You might consider moving some of this to `src/parse/asp/util.go`; that's a good place for generic utility functions.

At some point we could probably optimise it a bit too (like I think right now it basically walks the whole AST top-down, it could skip through the `[]*Statement` entries rapidly until it finds one at the right position)

Reply:
Let's have a chat about this over slack.